pulpcore/plugin/serializers/content.py

```diff
@@ -202,6 +202,8 @@ def deferred_validate(self, data):
                 artifact = Artifact.objects.get(
                     sha256=file.hashers["sha256"].hexdigest(), pulp_domain=get_domain_pk()
                 )
+                if not artifact.pulp_domain.get_storage().exists(artifact.file.name):
+                    raise Artifact.DoesNotExist
                 artifact.touch()
```

Comment:
I don't think, this is quite correct. The exception handling code here will try to save a new artifact and probably fail.

How about this?
```suggestion
                if not artifact.pulp_domain.get_storage().exists(artifact.file.name):
                    artifact.file=file
                    artifact.save()
                else:
                    artifact.touch()
```

I think we should be able to add a test for this codepath using the url on the create call.

Reply:
Artifact.save() doesn't do what we need - serializer.save() is what notices "file" and does the copy-tmp-to-perm. The change here is to notice that, in spite of the Artifact existing, the file does *not*, and to do the exception path. This is the path that is being tested by `test_reupload_damaged_artifact`. What am I missing?

Reply:
Thanks for the clarification. Serializer save actually works different from model save. But does that serializer not need to have the artifact assigned as instance to know it should do an update? I'm probably missing something too.

I thought the test you provided would exercise the change in the viewset code.
(As far as I understand the artifact saving code is doubled, because on atomic upload the whole file is know by the viewset and we prevent moving the bits over the wire tree times. On chunked upload, or download however the file is only available in deferred_validate.)

Reply:
> (As far as I understand the artifact saving code is doubled, because on atomic upload the whole file is know by the viewset and we prevent moving the bits over the wire tree times. On chunked upload, or download however the file is only available in deferred_validate.)

Ah! *That* is what I was missing. I'll redraft this and write up a test for that.

Reply:
Thank you, I think we want both paths to be covered anyway.